### PR TITLE
tools/ratbagd.py: sync with Piper

### PR DESCRIPTION
**Pull request by Hjdskes**

### CODE REVIEW

**Comment by Hjdskes** [comment]
You could do away with some maps in `tools/ratbagctl.in` if you want; there are equivalents in the bindings although they give slightly different results, i.e. "Doubleclick" versus "doubleclick".

**Comment by Hjdskes** [comment]
Oke I give up on the other failures; I should start writing my GSoC report.

**Comment by whot** [comment]
First issue I found is with the `IsActive` handling in the profile. The test selects a profile which sends `SetActive`. That triggers the cached dbus property update but does **not** unset `IsActive` on the currently active profile. When we then run the test we expect profile 2 to be active, but profile 0 is active too and because we search in order we find 0 first. That triggers the test failures because button mappings are different between the profiles.

This is when running 
`sudo ./build/ratbagctl.test TestRatbgagCtlButton.test_button_n_action_set_button`, I suspect all other tests have a variant of this issue too.

To fix this we'd have to trigger the main loop and wait for the dbus signal to arrive. No idea how this can be done in the test suite.

**Comment by Hjdskes** [comment]
That'd make sense, yea. Two ways to fix this:

1. Make ratbagctl work using the main loop, somehow, or
2. Diverge the bindings between ratbagctl and Piper. This isn't as far-fetched as it sounds as the bindings are really written for Piper with GObject properties and signals and so forth, so you can simplify a lot by making them diverge.

**Comment by bentiss** [comment]
> 1. Make ratbagctl work using the main loop, somehow, or
> 2. Diverge the bindings between ratbagctl and Piper. This isn't as far-fetched as it sounds as the bindings are really written for Piper with GObject properties and signals and so forth, so you can simplify a lot by making them diverge.

I'd rather 1. I still prefer having piper using a reliable ratbagd.py that has gone through the regression tests. Given that we already share the GObject code, it doesn't hurt to run the GLib mainloop. "How" is an other story.

**Comment by Hjdskes** [comment]
> I still prefer having piper using a reliable ratbagd.py that has gone through the regression tests.

Piper can always add its own tests if need be, where we then would also have tests for the signals of changed properties and such.

**Comment by Hjdskes** [comment]
I've been debugging the test failures and it seems that all of them but one are caused by this change:

```python
diff --git a/tools/ratbagd.py b/tools/ratbagd.py
index d9b29ac..a73251e 100644
--- a/tools/ratbagd.py
+++ b/tools/ratbagd.py
@@ -301,6 +301,11 @@ class RatbagdDevice(_RatbagdDBus):
     def __init__(self, object_path):
         _RatbagdDBus.__init__(self, "Device", object_path)
 
+        # FIXME: if we start adding and removing objects from this list,
+        # things will break!
+        result = self._get_dbus_property("Profiles")
+        self._profiles = [RatbagdProfile(objpath) for objpath in result]
+
     @GObject.Property
     def id(self):
         """The unique identifier of this device."""
@@ -324,11 +329,7 @@ class RatbagdDevice(_RatbagdDBus):
     @GObject.Property
     def profiles(self):
         """A list of RatbagdProfile objects provided by this device."""
-        profiles = []
-        result = self._get_dbus_property("Profiles")
-        if result is not None:
-            profiles = [RatbagdProfile(objpath) for objpath in result]
-        return profiles
+        return self._profiles
 
     def get_svg(self, theme):
         """Gets the full path to the SVG for the given theme, or the empty
```

In [this branch](https://github.com/Hjdskes/libratbag/commits/wip/sync) I have synced most of the changes from Piper to libratbag and there's only one test failure there, which I do suspect is due to a missing main loop. As soon as I add the above diff, I get 12 more test failures. The above diff, however, cannot be because of a missing main loop and must be something else.

**Comment by Hjdskes** [comment]
After merging the remainder, it's this little diff that causes the 12 test failures:

```diff
diff --git a/tools/ratbagd.py b/tools/ratbagd.py
index 542f083..ec601cf 100644
--- a/tools/ratbagd.py
+++ b/tools/ratbagd.py
@@ -339,11 +339,7 @@ class RatbagdDevice(_RatbagdDBus):
     @GObject.Property
     def profiles(self):
         """A list of RatbagdProfile objects provided by this device."""
-        profiles = []
-        result = self._get_dbus_property("Profiles")
-        if result is not None:
-            profiles = [RatbagdProfile(objpath) for objpath in result]
-        return profiles
+        return self._profiles
 
     @GObject.Property
     def active_profile(self):
```

I'm quitting for the night so if you beat me to debugging this one, feel free to take my branch and merge it :)

**Comment by Hjdskes** [comment]
Oke things work now in my branch; if you spawn ratbagd in a subprocess things magically start working. The downside is that it is slower and some tests don't work because

```
File "/home/jente/code/src/github.com/libratbag/libratbag/build/ratbagctl", line 1117, in print_button
    print("{}macro '{}'".format(header, str(b.macro)))
UnicodeEncodeError: 'ascii' codec can't encode character '\u2195' in position 45: ordinal not in range(128)
```

This is the ↕ arrow we're using to print macros; even if I specify an encoding / decoding of utf-8 I get this message. I'm not that familiar with `Popen` so I haven't solved this yet. dbus-daemon also takes quite a bit of CPU now, not sure where that's coming from. Now I'm really out for the night :P Sorry for the noise.

**Comment by whot** [comment]
> if you spawn ratbagd in a subprocess things magically start working. 

That's not quite right here: you're spawning ratbagctl as subprocess, ratbagd is always a subprocess.This means if you have 5 commands in a test, you spawn it 5 times (see: CPU usage). Each instance will get the updated values from ratbagd correctly because there's no buffering. but that's also where the slowdown comes from.

fwiw, this diff makes your branch pass all tests here (even the commented-out ones)
````diff --git a/tools/ratbagctl.test.in b/tools/ratbagctl.test.in
index 872340d..f7ae18a 100755
--- a/tools/ratbagctl.test.in
+++ b/tools/ratbagctl.test.in
@@ -50,8 +50,8 @@ class TestRatbagCtl(unittest.TestCase):
 
         (stdout, stderr) = ratbagctl.communicate()
         return (ratbagctl.returncode,
-                stdout.decode('ascii').rstrip('\n'),
-                stderr.decode('ascii').rstrip('\n'))
+                stdout.decode('utf-8').rstrip('\n'),
+                stderr.decode('utf-8').rstrip('\n'))
````

**Comment by Hjdskes** [comment]
> That's not quite right here: you're spawning ratbagctl as subprocess

Whoops. My excuse is it was late :)

> fwiw, this diff makes your branch pass all tests here (even the commented-out ones)

Weird, I had that but was still getting the above error. Let me try again soon.

Anyway, this is then probably not the right way to solve this problem but I haven't managed to start a main loop correctly yet. Will play/ask around more today.

**Comment by Hjdskes** [comment]
> > fwiw, this diff makes your branch pass all tests here (even the commented-out ones)
> Weird, I had that but was still getting the above error. Let me try again soon.

Nope, still doesn't work. Probably something local to my setup then. In any case, I'll be looking into the main loop again and will be reverting those changes.

**Comment by bentiss** [comment]
I spent some time today on this, given that this one prevents other PR to get in.

First, we are not forced to run the mainloop, we can just pull the events at the right time:
```
diff --git a/tools/ratbagctl.test.in b/tools/ratbagctl.test.in
index c3a71ae..e81fa0a 100755
--- a/tools/ratbagctl.test.in
+++ b/tools/ratbagctl.test.in
@@ -53,8 +53,8 @@ class TestRatbagCtl(unittest.TestCase):
 
         (stdout, stderr) = ratbagctl.communicate()
         return (ratbagctl.returncode,
-                stdout.decode('ascii').rstrip('\n'),
-                stderr.decode('ascii').rstrip('\n'))
+                stdout.decode('utf-8').rstrip('\n'),
+                stderr.decode('utf-8').rstrip('\n'))
 
     @classmethod
     def run_ratbagctl_fast(cls, params):
@@ -74,6 +74,7 @@ class TestRatbagCtl(unittest.TestCase):
         error = sys.stderr.getvalue()
         sys.stdout = stdout
         sys.stderr = stderr
+        toolbox.sync_dbus()
         return returncode, output.rstrip('\n'), error.rstrip('\n')
 
     @classmethod
diff --git a/tools/toolbox.py b/tools/toolbox.py
index cd826e3..9db96ce 100644
--- a/tools/toolbox.py
+++ b/tools/toolbox.py
@@ -27,6 +27,8 @@ import shutil
 import subprocess
 import sys
 
+from gi.repository import GLib
+
 # various constants
 RATBAGCTL_NAME = 'ratbagctl'
 RATBAGCTL_PATH = os.path.join('@MESON_BUILD_ROOT@', RATBAGCTL_NAME)
@@ -98,6 +100,13 @@ def terminate_ratbagd(ratbagd):
         ratbagd.kill()
     os.unlink(DBUS_CONF_PATH)
 
+
+def sync_dbus():
+    main_context = GLib.MainContext.default()
+    while main_context.pending():
+        main_context.iteration(False)
+
+
 ratbagctl = import_non_standard_path(RATBAGCTL_NAME, RATBAGCTL_PATH)
 
 from ratbagctl import open_ratbagd, get_parser
@@ -111,5 +120,6 @@ __all__ = [
     start_ratbagd,
     terminate_ratbagd,
     open_ratbagd,
+    sync_dbus,
     get_parser,
 ]
```

This allows to get the dbus signals when properties are changed.

I then tried to make `TestRatbagCtlProfile` test working (it fails at set active profile). It's a requirement for this to pass before other tests are run.

So, I realized we have an issue in ratbagd in the way we update the profiles on set_active():
```
diff --git a/ratbagd/ratbagd-profile.c b/ratbagd/ratbagd-profile.c
index 40cbd13..cf767c7 100644
--- a/ratbagd/ratbagd-profile.c
+++ b/ratbagd/ratbagd-profile.c
@@ -294,12 +294,15 @@ static int ratbagd_profile_set_active(sd_bus_message *m,
        if (r < 0)
                return r;
 
+       r = ratbag_profile_set_active(profile->lib_profile);
+       if (r < 0)
+               return r;
+
        ratbagd_for_each_profile_signal(sd_bus_message_get_bus(m),
                                        profile->device,
                                        ratbagd_profile_active_signal_cb);
 
-        return sd_bus_reply_method_return(m, "u",
-                                          ratbag_profile_set_active(profile->lib_profile));
+        return sd_bus_reply_method_return(m, "u", 0);
 }
 
 static int
```

We should update the values before sending the events that they are changed.

And finally, I think the ratbagd.py binding is messing up somewhere as there are no g-object event raised that I can find on property changed on profile set_active, meaning that we are basically never in a situation where the test suite will work.

Just to be clear, with the bindings from this PR, you can see than setting a profile active doesn't deactivate the previous one, leading to a list of _profiles[] where 2 profiles are active.

Just my 2 cents to help on the bug.

**Comment by whot** [comment]
I've pushed the fix for the `ratbagd_profile_set_active()` to master as 432d32b

**Comment by whot** [comment]
ok, found the problem, see #315 

@Hjdskes Check out https://github.com/whot/libratbag/tree/wip/sync-ratbagd, it has the commits almost ready to go, still fails with a button-macro related test case. but hopefully that one should be easy enough to fix :)

**Comment by Hjdskes** [comment]
@whot and @bentiss Thanks for the help, things are passing locally now so I expect they'll pass on CircleCI as well.

> And finally, I think the ratbagd.py binding is messing up somewhere as there are no g-object event raised that I can find on property changed on profile set_active, meaning that we are basically never in a situation where the test suite will work.
> 
> Just to be clear, with the bindings from this PR, you can see than setting a profile active doesn't deactivate the previous one, leading to a list of _profiles[] where 2 profiles are active.

I think that'll be fixed by https://github.com/libratbag/piper/issues/167

**Comment by Hjdskes** [comment]
Sync of https://github.com/libratbag/piper/pull/168 is ready in [`sync-ratbagd-2](https://github.com/Hjdskes/libratbag/tree/sync-ratbagd-2), feel free to cherry-pick once this PR and https://github.com/libratbag/piper/pull/168 are in. Once that is done I'll sync the profile name changes back into Piper.
